### ISSUE TITLE
JAX asin accuracy and correctness

### ISSUE THREAD

**Issue by pearu** [pearu/functional_algorithms#38]
```
  -inf    11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  -3e35   11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  -3e32   11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  -3e29   11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  -3e26   11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  -3e23   11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  -3e20   11111112111111111111111111111111111111111111111111111111111111111111111111111111111111111111211111111
  -3e17   11111112122111111111111111111111111111111111111111111111111111111111111111111111111111111121211111111
  -3e14   11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  -3e11   111111111111111=1111111=2====================================================211111111111111111111111
  -2e8    11111111111111111=11122222=================================================22222111111111111111111111
  -2e5    11111111111111==111=11212211211111111111111111111111111111111111111111111112222111111==11111111111111
  -2e2    11111111111111====11111122233321222112222222222211112121222221111122122223232111111====11111111111111
  -2e-1   11111111111111======111112333331222112221222122222E222222221222222222233333311111======11111111111111
  -2e-4   11111111111111======111112222211111111111111111111E1111111111111111111122222111========11111111111111
  -2e-7   11111111111111======11111121111===================E===================1111111==========11111111111111
  -2e-10  11111111111111======11111121111===================E===================1111111==========11111111111111
  -2e-13  11111111111111======11111121111===================E===================1111111==========11111111111111
  -2e-16  11111111111111======11111121111===================E===================1111111==========11111111111111
  -2e-19  11111111111111======11111121111===================E===================1111111==========11111111111111
  -2e-22  11111111111111======11111121111===================E===================1111111==========11111111111111
  -2e-25  11111111111111======11111121111===================E===================1111111==========11111111111111
  -2e-28  11111111111111======11111121111===================E===================1111111==========11111111111111
  -2e-31  11111111111111======11111121111===================E===================1111111==========11111111111111
  -1e-34  11111111111111======11111121111===================E===================1111111==========11111111111111
  -1e-37  EEEEEEEEEEEEEEEEEEEEEEEEEE21111===================E===================11111EEEEEEEEEEEEEEEEEEEEEEEEEE
  1e-37   11111111111111======11111121111===================E===================1111111==========11111111111111
  1e-34   11111111111111======11111121111===================E===================1111111==========11111111111111
  2e-31   11111111111111======11111121111===================E===================1111111==========11111111111111
  2e-28   11111111111111======11111121111===================E===================1111111==========11111111111111
  2e-25   11111111111111======11111121111===================E===================1111111==========11111111111111
  2e-22   11111111111111======11111121111===================E===================1111111==========11111111111111
  2e-19   11111111111111======11111121111===================E===================1111111==========11111111111111
  2e-16   11111111111111======11111121111===================E===================1111111==========11111111111111
  2e-13   11111111111111======11111121111===================E===================1111111==========11111111111111
  2e-10   11111111111111======11111121111===================E===================1111111==========11111111111111
  2e-7    11111111111111======11111121111===================E===================111122111========11111111111111
  2e-4    11111111111111======111113333332222222222222222222E222222222222222222233333211111======11111111111111
  2e-1    11111111111111====11111122332321111221111221111221112222122211221222122223332111111====11111111111111
  3e2     11111111111111==111=11223222222222222222222222221===122222222222222222222222322111111==11111111111111
  3e5     11111111111111111=1123222222222=======================================2222222222111111111111111111111
  3e8     111111111111111=111122211===================================================1=1111111=111111111111111
  3e11    11111111112111111111111111111111111111111111111111111111111111111111111111111111111111111221211111111
  3e14    11111111112111111111111111111111111111111111111111111111111111111111111111111111111111111121211111111
  4e17    11111112111111111111111111111111111111111111111111111111111111111111111111111111111111111111211111111
  4e20    11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  4e23    11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  4e26    11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  4e29    11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  5e32    11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  5e35    11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
          -inf  -3e29 -2e20 -1e11 -1e2  -8e-8 -6e-1 -4e-2 -3e-3 1e-33 1e-24 2e-15 3e-6  4e3   5e12  8e21  1e31
```

| ULP-difference |                              z |             jax:asin(z) |          mpmath:asin(z) |           numpy:asin(z) |              fa:asin(z) |
| -------------- | ------------------------------ | ----------------------- | ----------------------- | ----------------------- | ----------------------- |
|     4294967296 |            (-inf-5.88902e-39j) |       (-1.5707964+infj) |       (-1.5707964-infj) |       (-1.5707964-infj) |       (-1.5707964-infj) |
|     4294967296 |             (inf-5.88902e-39j) |        (1.5707964+infj) |        (1.5707964-infj) |        (1.5707964-infj) |        (1.5707964-infj) |
|     2237116610 | (-1.03440016e+37-5.88902e-39j) | (-1.5707964+85.922615j) | (-1.5707964-85.922615j) | (-1.5707964-85.922615j) | (-1.5707964-85.922615j) |
|     2237115934 |   (1.0317341e+37-5.88902e-39j) |   (1.5707964+85.92004j) |   (1.5707964-85.92004j) |   (1.5707964-85.92004j) |   (1.5707964-85.92004j) |
|     2236200639 |  (-3.1419312e+35-5.88902e-39j) |  (-1.5707964+82.42846j) |  (-1.5707964-82.42847j) |  (-1.5707964-82.42846j) |  (-1.5707964-82.42846j) |
|     2236199245 |   (3.1252682e+35-5.88902e-39j) |   (1.5707964+82.42314j) |   (1.5707964-82.42315j) |   (1.5707964-82.42314j) |   (1.5707964-82.42314j) |
|     2235284450 |   (-9.535506e+33-5.88902e-39j) |  (-1.5707964+78.93348j) |  (-1.5707964-78.93348j) |  (-1.5707964-78.93348j) |  (-1.5707964-78.93348j) |
|     2235282294 |   (9.4573977e+33-5.88902e-39j) |  (1.5707964+78.925255j) |  (1.5707964-78.925255j) |  (1.5707964-78.925255j) |  (1.5707964-78.925255j) |
|     2234368029 |  (-2.8913991e+32-5.88902e-39j) |  (-1.5707964+75.43761j) | (-1.5707964-75.437614j) |  (-1.5707964-75.43761j) |  (-1.5707964-75.43761j) |
|     2234365062 |   (2.8588539e+32-5.88902e-39j) |   (1.5707964+75.42629j) |   (1.5707964-75.42629j) |   (1.5707964-75.42629j) |   (1.5707964-75.42629j) |
|     2233451363 |  (-8.7592267e+30-5.88902e-39j) |   (-1.5707964+71.9408j) |  (-1.5707964-71.94081j) |   (-1.5707964-71.9408j) |   (-1.5707964-71.9408j) |
|     2233447530 |   (8.6320967e+30-5.88902e-39j) |  (1.5707964+71.926186j) |  (1.5707964-71.926186j) |  (1.5707964-71.926186j) |  (1.5707964-71.926186j) |
|     2232534436 |  (-2.6508847e+29-5.88902e-39j) |  (-1.5707964+68.44301j) |  (-1.5707964-68.44301j) |  (-1.5707964-68.44301j) |  (-1.5707964-68.44301j) |
|     2232529678 |    (2.603211e+29-5.88902e-39j) |   (1.5707964+68.42486j) |   (1.5707964-68.42486j) |   (1.5707964-68.42486j) |   (1.5707964-68.42486j) |
|     2231617230 |  (-8.0140974e+27-5.88902e-39j) | (-1.5707964+64.944145j) | (-1.5707964-64.944145j) | (-1.5707964-64.944145j) | (-1.5707964-64.944145j) |
|     2231611482 |   (7.8402865e+27-5.88902e-39j) |   (1.5707964+64.92222j) |   (1.5707964-64.92222j) |   (1.5707964-64.92222j) |   (1.5707964-64.92222j) |
|     2230029726 |  (-2.4200562e+26-5.88902e-39j) |  (-1.5707964+61.44415j) |  (-1.5707964-61.44415j) |  (-1.5707964-61.44415j) |  (-1.5707964-61.44415j) |
|     2230016103 |   (2.3579809e+26-5.88902e-39j) |  (1.5707964+61.418163j) |  (1.5707964-61.418167j) |  (1.5707964-61.418163j) |  (1.5707964-61.418163j) |
|     2228194083 |  (-7.2990844e+24-5.88902e-39j) | (-1.5707964+57.942936j) |  (-1.5707964-57.94294j) | (-1.5707964-57.942936j) | (-1.5707964-57.942936j) |
|     2228178168 |   (7.0808513e+24-5.88902e-39j) |  (1.5707964+57.912582j) |  (1.5707964-57.912582j) |  (1.5707964-57.912582j) |  (1.5707964-57.912582j) |
|            ... |                            ... |                     ... |                     ... |                     ... |                     ... |

**Comment by pearu** [comment]
The errors above occur when the input to arcsin uses subnormals.

Interestingly, the result depends on the device:
```python
>>> arr = numpy.array([complex(-numpy.inf, -i) for i in [1e-37, 1e-38, 1e-45, 1e-46]], dtype=numpy.complex64)
>>> print(numpy.arcsin(arr).imag)  # expected result
[-inf -inf -inf -inf]
>>> print(jax.numpy.arcsin(jax.numpy.array(arr, device=jax.devices('cpu')[0])).imag)
[-inf  inf  inf  inf]
>>> print(jax.numpy.arcsin(jax.numpy.array(arr, device=jax.devices('cuda')[0])).imag)
[-inf -inf -inf  inf]
```

**Comment by pearu** [comment]
When disabling tests with subnormals, we'll have:
```
 -inf    11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  -3e35   11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  -3e32   11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  -3e29   11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  -3e26   11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  -3e23   11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  -3e20   11111111211111111111111111111111111111111111111111111111111111111111111111111111111111111111211111111
  -3e17   1=111111221111111111111111111111111111111111111111111111111111111111111111111111111111111111211111111
  -3e14   1=111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  -3e11   1=1111111111111=11111111======================================================2111111=111111111111111
  -3e8    1=111111111111111=11122222111111111111111111111111111111111111111111111111122222111111111111111111111
  -3e5    1=111111111111==111111223222222222222222222222222===22222222222222222222222222211=111==11111111111111
  -3e2    1=111111111111====11111122332212111121111112211111111112211112111111221223332211111====11111111111111
  -3e-1   1=111111111111======1111123333!22222222222222222222222222222222222222233333211111======11111111111111
  -3e-4   1=111111111111=======1111222111111111111111111111111111111111111111111111222111========11111111111111
  -3e-7   1=111111111111=======1=11112111=======================================11111111=========11111111111111
  -3e-10  1=111111111111=======1=11112111=======================================11111111=========11111111111111
  -3e-13  1=111111111111=======1=11112111=======================================11111111=========11111111111111
  -3e-16  1=111111111111=======1=11112111=======================================11111111=========11111111111111
  -3e-19  1=111111111111=======1=11112111=======================================11111111=========11111111111111
  -3e-22  1=111111111111=======1=11112111=======================================11111111=========11111111111111
  -3e-25  1=111111111111=======1=11112111=======================================11111111=========11111111111111
  -3e-28  1=111111111111=======1=11112111=======================================11111111=========11111111111111
  -3e-31  1=111111111111=======1=11112111=======================================11111111=========11111111111111
  -3e-34  1=111111111111=======1=11112111=======================================11111111=========11111111111111
  -3e-37  1=111111111111=======1=11112111=======================================11111111=========11111111111111
  3e-37   1=111111111111=======1=11112111=======================================11111111=========11111111111111
  3e-34   1=111111111111=======1=11112111=======================================11111111=========11111111111111
  3e-31   1=111111111111=======1=11112111=======================================11111111=========11111111111111
  3e-28   1=111111111111=======1=11112111=======================================11111111=========11111111111111
  3e-25   1=111111111111=======1=11112111=======================================11111111=========11111111111111
  3e-22   1=111111111111=======1=11112111=======================================11111111=========11111111111111
  3e-19   1=111111111111=======1=11112111=======================================11111111=========11111111111111
  3e-16   1=111111111111=======1=11112111=======================================11111111=========11111111111111
  3e-13   1=111111111111=======1=11112111=======================================11111111=========11111111111111
  3e-10   1=111111111111=======1=1111211111111111111111111111111111111111111111111111111=========11111111111111
  4e-7    1=111111111111=======1111212222111111111111111111111111111111111111111222121111========11111111111111
  4e-4    1=111111111111======1111133332222222222222222222222222222222222222222223333311111======11111111111111
  4e-1    1=111111111111====11111122322232212222122222222222222222222222222222222222232111111====11111111111111
  4e2     1=111111111111==111=11122221111111111111111111111===11111111111111111111112222211=111==11111111111111
  4e5     1=111111111111111=1122111111111=======================================1111111121111111111111111111111
  4e8     1=1111111111111=11111111=======================================================1111111111111111111111
  4e11    1=111111222111111111111111111111111111111111111111111111111111111111111111111111111111111211211111111
  4e14    1=111111221111111111111111111111111111111111111111111111111111111111111111111111111111111111211111111
  4e17    11111111211111111111111111111111111111111111111111111111111111111111111111111111111111111111211111111
  4e20    11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  5e23    11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  5e26    11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  5e29    11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  5e32    11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
  5e35    11111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111
          -inf  -3e29 -2e20 -2e11 -1e2  -1e-7 -1e-1 -8e-2 -6e-3 2e-33 3e-24 3e-15 4e-6  5e3   7e12  9e21  1e31
```

| ULP-difference |                           z |                  jax:asin(z) |               mpmath:asin(z) |                numpy:asin(z) |                   fa:asin(z) |
| -------------- | --------------------------- | ---------------------------- | ---------------------------- | ---------------------------- | ---------------------------- |
|              4 | (-5.818689e-08-0.03941552j) | (-5.8141758e-08-0.03940532j) | (-5.8141744e-08-0.03940532j) | (-5.8141744e-08-0.03940532j) | (-5.8141758e-08-0.03940532j) |

**Comment by pearu** [comment]
When disabling tests with subnormals, we'll have (on CUDA): all tests pass.

**Comment by pearu** [comment]
Closing via https://github.com/pearu/functional_algorithms/pull/56 that disables subnormal samples for JAX
